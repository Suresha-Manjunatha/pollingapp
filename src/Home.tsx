/** @format */

import { useNavigation } from "@react-navigation/native";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import axios from "axios";
import React, { useEffect, useState } from "react";
import { FlatList, View, Text, TouchableOpacity } from "react-native";
import { ScreenStyles } from "./styles";

const Home = () => {
  const [data, setData] = useState<any[]>([]);
  const [page, setPage] = useState<number>(1);
  const navigation = useNavigation<NativeStackNavigationProp<any>>();
  let interval: NodeJS.Timer;
  const [loading, setloading] = useState<boolean>(false);

  const handelPage = () => {
    setPage((p: any) => p + 1);
  };

  const fetchData = async () => {
    try {
      setloading(true);
      const res = await axios.get(
        `https://hn.algolia.com/api/v1/search_by_date?query=story&page=${page}`
      );
      //   console.log(res.data.hits);
      setData([...data, ...res.data.hits]);
    } catch (e) {
      console.log("error");
    } finally {
      setloading(false);
    }
  };

  useEffect(() => {
    fetchData();
    console.log(page);
  }, [page]);

  useEffect(() => {
    interval = setInterval(() => handelPage(), 10000);
    return () => clearInterval(interval);
  }, []);

  const handleScroll = () => {
    if (
      window.scrollY + window.innerHeight >=
      document.documentElement.scrollHeight
    ) {
      !loading && handelPage();
    }
  };

  useEffect(() => {
    window.addEventListener("scroll", handleScroll);

    return () => window.removeEventListener("scroll", handleScroll);
  }, []);

  return (
    <View testID="home" style={ScreenStyles.screen}>
      <FlatList
        horizontal={false}
        testID="home-list"
        data={data}
        // keyExractor={(item:any)=>item.objectID}
        renderItem={({ item, index }) => {
          return (
            <TouchableOpacity
              testID={`test-${item.index}`}
              style={ScreenStyles.item}
              key={item.objectID}
              onPress={() => {
                navigation.navigate("RAW JSON", {
                  item: JSON.stringify(item, null, 4),
                });
              }}
            >
              <Text style={ScreenStyles.itemText1}>
                <Text style={ScreenStyles.title}>Title: </Text>
                {item.title ? (
                  item.title
                ) : item.story_title ? (
                  item.story_title
                ) : (
                  <i>data not found</i>
                )}
              </Text>
              <Text style={ScreenStyles.itemText1}>
                <Text style={ScreenStyles.title}>URL: </Text>
                {item.url ? (
                  <a href={item.url}>{item.url}</a>
                ) : item.story_url ? (
                  <a href={item.story_url}>{item.story_url}</a>
                ) : (
                  <i>data not found</i>
                )}
              </Text>
              <Text style={ScreenStyles.itemText1}>
                <Text style={ScreenStyles.title}>Created_at: </Text>
                {item.created_at}
              </Text>

              <Text style={ScreenStyles.itemText1}>
                <Text style={ScreenStyles.title}>Author: </Text>
                {item.author}
              </Text>
            </TouchableOpacity>
          );
        }}
        onEndReachedThreshold={0.5}
        // onEndReached={handelPage}
        // ListFooterComponent={() => {
        //   if (loading) {
        //     return <ActivityIndicator />;
        //   } else return <View />;
        // }}
      />
    </View>
  );
};

export default Home;
