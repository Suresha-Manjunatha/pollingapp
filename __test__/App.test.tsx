import Home from "../src/Home";
import { render } from "@testing-library/react-native";

jest.mock("@react-navigation/native", () => {
  return {
    useNavigation: () => ({
      navigate: jest.fn(),
    }),
  };
});

describe("home component test", () => {
  jest.useFakeTimers();
  it("take snap shot", () => {
    render(<Home />);
  });
});
