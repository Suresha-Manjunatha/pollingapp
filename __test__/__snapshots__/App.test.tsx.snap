// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home component test take snap shot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "marginTop": 20,
      "paddingTop": 20,
    }
  }
  testID="home"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "color": "lightblue",
        "display": "flex",
        "flexDirection": "row",
        "fontSize": 22,
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "flex": 1,
          "height": 150,
          "overflow": "hidden",
          "padding": 5,
          "width": "20%",
        }
      }
    >
      Title
    </Text>
    <Text
      style={
        Object {
          "flex": 1,
          "flexWrap": "wrap",
          "height": 150,
          "padding": 5,
          "width": "30%",
        }
      }
    >
      URL
    </Text>
    <Text
      style={
        Object {
          "flex": 1,
          "flexWrap": "wrap",
          "height": 150,
          "padding": 5,
          "width": "30%",
        }
      }
    >
      Created_At
    </Text>
    <Text
      style={
        Object {
          "flex": 1,
          "height": 150,
          "overflow": "hidden",
          "padding": 5,
          "width": "20%",
        }
      }
    >
      Author
    </Text>
  </View>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="home-list"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;
